File: _05_Merge_sort.cpp
#include<iostream>
using namespace std;

void printArray(int array[], int n);
void merge(int array[],int lb,int mid, int ub);
//Merge sort : 
void mergeSort(int array[], int lb, int ub){
    if(lb<ub){
        int mid = (lb+ub)/2;
        mergeSort(array,lb,mid);
        mergeSort(array,mid+1,ub);
        merge(array,lb,mid,ub);
    }
}

//Merge : 
void merge(int array[], int lb, int mid, int ub){
    int i = lb;
    int j = mid+1;
    int k = lb;

    int helperArray[lb+ub];

    while (i<=mid && j<= ub)
    {
        if(array[i]<=array[j]){
            helperArray[k] = array[i];
            k++,i++;
        }else{
            helperArray[k] = array[j];
            k++,j++;
        }
    }

    //if left:
    if(i>mid){
        while (j<=ub)
        {
            helperArray[k] = array[j];
            k++,j++;
        } 
    }else{
        while (i<=mid)
        {
            helperArray[k] = array[i];
            k++,i++;
        }
        
    }
    

    //Copying helper to our main array : 
    for (int i = lb; i <= ub; i++)
    {
        array[i] = helperArray[i];
    }
    
}


//printing : 
void printArray(int array[], int n){
    for (int i = 0; i < n; i++)
    {
        cout << array[i]<< " ";
    }cout << endl;
}

int main(){
    int n;
    cout << "Enter the size of the array : \n";
    cin >> n;
    int array[n];
    cout << "Enter the array : \n";
    for(int i = 0; i < n; i++)
    {
        cin >> array[i];
    }
    int lb = 0, ub = n-1;
    mergeSort(array,lb,ub);

    cout << "\nAfter Merge Sorting :\n";
    printArray(array,n);
    
}

File: _05_Merge_sort.java
import java.util.Scanner;

public class _05_Merge_sort {
    public static void main(String[] args) {
        int[] array = input();
        int n = array.length;
        int lb = 0, ub = n-1;

        mergeSort(array,lb,ub);
        System.out.println("\nAfter Merge Sorting : ");
        printArray(array,n);
    }
//Merge Sort : 
    public static void mergeSort(int[] array, int lb, int ub){
        if(lb<ub){
            int mid = (lb+ub)/2;
            mergeSort(array, lb, mid);
            mergeSort(array, mid+1, ub);
            merge(array, lb, mid, ub);
        }
    }

//Merge : 
    public static void merge(int[] array, int lb, int mid, int ub){
        int i = lb;
        int j = mid+1;
        int k = 0;
        
        int helperArray[] = new int[(ub - lb + 1)];

        while (i<=mid && j<=ub) {
            if(array[i]<=array[j]){
                helperArray[k] = array[i];
                k++;
                i++;
            }else{
                helperArray[k] = array[j];
                k++;
                j++;
            }
        }

        if(i>mid){
            while (j<=ub){
                helperArray[k] = array[j];
                k++;
                j++;
            }
        }else{
            while (i<=mid) {
                helperArray[k] = array[i];
                k++;
                i++;
            }
        }

        for(int m = 0; m < helperArray.length; m++){
            array[lb+m] = helperArray[m];
        }
    }


//Taking input : 
    public static int[] input(){
        int n;
        System.out.println("Enter the size of the array : ");
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        int[] array = new int[n];
        System.out.println("\nEnter the array : ");
        for(int i = 0; i < n; i++){
            array[i] = scanner.nextInt();
        }
        scanner.close();
        return array;
    }

    //Print array : 
    public static void printArray(int[] array, int n){
        for(int i = 0; i < n; i++){
            System.out.print(array[i]+" ");
        }System.out.println();
    }
}
